app/src/main/java/org/p2p/wallet/common/ui/widget/EarnWidget.kt

```diff
@@ -0,0 +1,183 @@
+                buttonEarn.isEnabled = true // TODO remove after tests and real integration
+            }
+            is EarnWidgetState.DepositFoundsFailed -> {
+                makeAlignCenterContent()
+                textViewEarnTitle.setText(R.string.earn_widget_deposit_failed_title)
+                textViewEarnMessage.setText(R.string.earn_widget_deposit_failed_message)
+                buttonEarn.setText(R.string.earn_widget_deposit_failed_button)
+            }
+            is EarnWidgetState.Error -> {
+                makeAlignCenterContent()
+                textViewEarnTitle.setText(R.string.earn_widget_error_title)
+                textViewEarnMessage.setText(state.messageTextRes)
+                buttonEarn.setText(state.buttonTextRes)
+            }
+        }
+    }
+
+    private fun makeAlignStartContent() = with(binding) {
+        textViewEarnTitle.apply {
+            textAlignment = TEXT_ALIGNMENT_TEXT_START
+            setTextAppearance(R.style.UiKit_TextAppearance_Regular_Text3)
+        }
+        val horizontalPadding = SMALL_BUTTON_TEXT_PADDING_DP.toPx()
+        buttonEarn.apply {
+            textAlignment = TEXT_ALIGNMENT_TEXT_START
+            setPadding(horizontalPadding, 0, horizontalPadding, 0)
+        }
+    }
+
+    private fun makeAlignCenterContent() = with(binding) {
+        textViewEarnTitle.apply {
+            textAlignment = TEXT_ALIGNMENT_CENTER
+            setTextAppearance(R.style.UiKit_TextAppearance_SemiBold_Text1)
+        }
+        val horizontalPadding = DEFAULT_BUTTON_TEXT_PADDING_DP.toPx()
+        buttonEarn.apply {
+            textAlignment = TEXT_ALIGNMENT_CENTER
+            setPadding(horizontalPadding, 0, horizontalPadding, 0)
+        }
+    }
+
+    fun setDepositTokens(tokens: List<String>) {
+        val container = binding.viewTokenContainer
+        if (container.childCount < tokens.size) {
+            for (i in container.childCount until tokens.size) {
+                val binding = ItemDepositTokenBinding.inflate(
+                    LayoutInflater.from(context), container, true
+                ).apply {
+                    val params = tokenImageView.layoutParams as MarginLayoutParams
+                    params.marginEnd = (TOKEN_MARGIN_END_DP * i).toPx()
+                    tokenImageView.layoutParams = params
+                }
+                val holder = DepositTokenViewHolder(binding, glideManager)
+                binding.root.tag = holder
+            }
+        } else if (container.childCount > tokens.size) {
+            if (tokens.isEmpty()) {
+                removeAllViews()
+            } else {
+                container.removeViewsInLayout(
+                    tokens.size, container.childCount - tokens.size
+                )
+            }
+        }
+        for (i in 0 until container.childCount) {
+            (container.getChildAt(i)?.tag as? DepositTokenViewHolder)?.apply {
+                bind(tokens[i])
+            }
```

Comment:
Why so complex?
Can we make it in xml or smth?
It would be hard to support

Reply:
nothing to support here
this logic is just to make viewHolders and use views to reuse an bing them like in RecyclerView
not to remove and add any time
because i do not know how often will we change deposit tokens


we also can't use xml because it could be 1 or 2 or 20 views here